app/controllers/Giraffe.scala

```diff
@@ -148,40 +154,71 @@ class Giraffe(paymentServices: PaymentServices) extends Controller with Redirect
     ), social, countryGroup, charge))
   }
 
-  def pay = NoCacheAction.async { implicit request =>
+  def pay = NoCacheAction.async(parse.form(supportForm)) { implicit request =>
+
+    val form = request.body
+
     val stripe = paymentServices.stripeServiceFor(request)
+    val idUser = IdentityUser.fromRequest(request).map(_.id)
+
+    val countryGroup = form.currency match {
+      case USD => US
+      case AUD => Australia
+      case EUR => Europe
+      case _ => UK
+    }
 
-    supportForm.bindFromRequest().fold[Future[Result]]({ withErrors =>
-      Future.successful(BadRequest(JsArray(withErrors.errors.map(k => JsString(k.key)))))
-    }, { f =>
-      val metadata = Map(
-        "marketing-opt-in" -> f.marketing.toString,
-        "email" -> f.email,
-        "name" -> f.name,
-        "abTests" -> Json.toJson(f.abTests).toString,
-        "ophanId" -> f.ophanId,
-        "cmp" -> f.cmp.mkString,
-        "intcmp" -> f.intcmp.mkString
-      ) ++ f.postcode.map("postcode" -> _)
-      // Note that '.. * 100' will not work for Yen and other currencies! https://stripe.com/docs/api#charge_object-amount
-      val amountInSmallestCurrencyUnit = (f.amount * 100).toInt
-      val maxAmountInSmallestCurrencyUnit = MaxAmount.forCurrency(f.currency) * 100
-      val res = stripe.Charge.create(min(maxAmountInSmallestCurrencyUnit, amountInSmallestCurrencyUnit), f.currency, f.email, "Your contribution", f.token, metadata)
-
-      val redirect = f.currency match {
-        case USD => routes.Giraffe.thanks(US).url
-        case AUD => routes.Giraffe.thanks(Australia).url
-        case EUR => routes.Giraffe.thanks(Europe).url
-        case _ => routes.Giraffe.thanks(UK).url
-      }
-
-      res.map { charge =>
-        Ok(Json.obj("redirect" -> redirect))
-          .withSession(chargeId -> charge.id)
-      }.recover {
-        case e: Stripe.Error => BadRequest(Json.toJson(e))
-      }
-    })
+    val variant = Test.getContributePageVariant(countryGroup, Test.testIdFor(request), request)
+
+    val contributionId = UUID.randomUUID()
```

Comment:
The reason it exists is to link the meta-data of a transaction with a transaction as they're stored in a different table.
I could make a contributionId type but I'm not sure we'd gain much from it